File: server/src/services/worktree.service.ts
import { basename, join, dirname } from 'path'
import type { Worktree, Agent } from '@claude-manager/shared'
import { WorktreeRepository } from '../db/repositories/worktree.repository.js'
import { AgentRepository } from '../db/repositories/agent.repository.js'
import { WorkspaceRepository } from '../db/repositories/workspace.repository.js'
import { NotFoundError, ConflictError, ValidationError } from '../utils/errors.js'
import { GitService } from './git.service.js'
import { logger } from '../utils/logger.js'

export interface WorktreeWithAgents extends Worktree {
  agents: Agent[]
  previousAgents: Agent[]
}

export interface CreateWorktreeOptions {
  workspaceId: string
  name: string
  branch: string
  createBranch: boolean
}

export class WorktreeService {
  constructor(
    private readonly worktreeRepo: WorktreeRepository,
    private readonly agentRepo: AgentRepository,
    private readonly workspaceRepo: WorkspaceRepository
  ) {}

  async createWorktree(options: CreateWorktreeOptions): Promise<Worktree> {
    const workspace = this.workspaceRepo.findById(options.workspaceId)
    if (!workspace) {
      throw new NotFoundError('Workspace', options.workspaceId)
    }

    // Check for name conflict
    if (this.worktreeRepo.existsWithNameInWorkspace(options.workspaceId, options.name)) {
      throw new ConflictError(`Worktree with name '${options.name}' already exists`)
    }

    // Generate worktree path - sibling to main repository
    const worktreePath = join(
      dirname(workspace.path),
      `${basename(workspace.path)}-${options.name}`
    )

    // Check if path already exists in another worktree
    if (this.worktreeRepo.findByPath(worktreePath)) {
      throw new ConflictError(`Worktree already exists at path: ${worktreePath}`)
    }

    // Create git worktree
    const gitService = new GitService(workspace.path)
    await gitService.addWorktree(worktreePath, options.branch, options.createBranch)

    // Create database record
    const worktree = this.worktreeRepo.create({
      workspaceId: options.workspaceId,
      name: options.name,
      branch: options.branch,
      path: worktreePath,
      isMain: false,
    })

    // Update workspace count
    this.workspaceRepo.incrementWorktreeCount(options.workspaceId)

    logger.info(
      { worktreeId: worktree.id, workspaceId: options.workspaceId, path: worktreePath },
      'Worktree created'
    )

    const { worktreeRowToApi } = await import('@claude-manager/shared')
    return worktreeRowToApi(worktree)
  }

  async getWorktreeById(id: string): Promise<Worktree | null> {
    const row = this.worktreeRepo.findById(id)
    if (!row) return null

    const { worktreeRowToApi } = await import('@claude-manager/shared')
    return worktreeRowToApi(row)
  }

  async getWorktreeWithAgents(id: string): Promise<WorktreeWithAgents | null> {
    const worktree = this.worktreeRepo.findById(id)
    if (!worktree) return null

    const { worktreeRowToApi, agentRowToApi } = await import('@claude-manager/shared')

    const activeAgents = this.agentRepo.findByWorktreeId(id, false)
    const deletedAgents = this.agentRepo.findDeletedByWorktreeId(id)

    return {
      ...worktreeRowToApi(worktree),
      agents: activeAgents.map(agentRowToApi),
      previousAgents: deletedAgents.map(agentRowToApi),
    }
  }

  async listWorktreesByWorkspace(workspaceId: string): Promise<Worktree[]> {
    const workspace = this.workspaceRepo.findById(workspaceId)
    if (!workspace) {
      throw new NotFoundError('Workspace', workspaceId)
    }

    const rows = this.worktreeRepo.findByWorkspaceId(workspaceId)
    const { worktreeRowToApi } = await import('@claude-manager/shared')
    return rows.map(worktreeRowToApi)
  }

  async updateWorktree(
    id: string,
    updates: { name?: string; sortMode?: 'free' | 'status' | 'name'; order?: number }
  ): Promise<Worktree> {
    const existing = this.worktreeRepo.findById(id)
    if (!existing) {
      throw new NotFoundError('Worktree', id)
    }

    // Check for name conflict if renaming
    if (
      updates.name &&
      this.worktreeRepo.existsWithNameInWorkspace(existing.workspace_id, updates.name, id)
    ) {
      throw new ConflictError(`Worktree with name '${updates.name}' already exists`)
    }

    const updated = this.worktreeRepo.update(id, {
      name: updates.name,
      sortMode: updates.sortMode,
      displayOrder: updates.order,
    })

    const { worktreeRowToApi } = await import('@claude-manager/shared')
    return worktreeRowToApi(updated!)
  }

  async deleteWorktree(id: string, force = false): Promise<void> {
    const worktree = this.worktreeRepo.findById(id)
    if (!worktree) {
      throw new NotFoundError('Worktree', id)
    }

    // Cannot delete main worktree
    if (worktree.is_main) {
      throw new ValidationError('Cannot delete the main worktree')
    }

    const workspace = this.workspaceRepo.findById(worktree.workspace_id)
    if (!workspace) {
      throw new NotFoundError('Workspace', worktree.workspace_id)
    }

    // Remove git worktree
    const gitService = new GitService(workspace.path)
    await gitService.removeWorktree(worktree.path, force)

    // Delete from database (cascades to agents)
    this.worktreeRepo.delete(id)

    // Update workspace count
    this.workspaceRepo.decrementWorktreeCount(worktree.workspace_id)

    logger.info({ worktreeId: id, workspaceId: worktree.workspace_id }, 'Worktree deleted')
  }

  async checkoutBranch(id: string, branch: string, createBranch = false): Promise<Worktree> {
    const worktree = this.worktreeRepo.findById(id)
    if (!worktree) {
      throw new NotFoundError('Worktree', id)
    }

    // Use git service with the worktree path
    const gitService = new GitService(worktree.path)
    await gitService.checkout(branch, createBranch)

    // Update database
    const updated = this.worktreeRepo.update(id, { branch })

    logger.info({ worktreeId: id, branch, createBranch }, 'Branch checked out')

    const { worktreeRowToApi } = await import('@claude-manager/shared')
    return worktreeRowToApi(updated!)
  }

  async reorderWorktrees(workspaceId: string, worktreeIds: string[]): Promise<void> {
    const workspace = this.workspaceRepo.findById(workspaceId)
    if (!workspace) {
      throw new NotFoundError('Workspace', workspaceId)
    }

    // Validate all IDs belong to this workspace
    const existingWorktrees = this.worktreeRepo.findByWorkspaceId(workspaceId)
    const existingIds = new Set(existingWorktrees.map((w) => w.id))

    for (const id of worktreeIds) {
      if (!existingIds.has(id)) {
        throw new ValidationError(`Worktree '${id}' does not belong to workspace`)
      }
    }

    this.worktreeRepo.reorder(workspaceId, worktreeIds)
  }

  async getGitStatus(
    id: string
  ): Promise<{ isClean: boolean; modified: string[]; staged: string[]; untracked: string[] }> {
    const worktree = this.worktreeRepo.findById(id)
    if (!worktree) {
      throw new NotFoundError('Worktree', id)
    }

    const gitService = new GitService(worktree.path)
    return gitService.getStatus()
  }

  async listBranches(id: string): Promise<{ local: string[]; remote: string[]; current: string }> {
    const worktree = this.worktreeRepo.findById(id)
    if (!worktree) {
      throw new NotFoundError('Worktree', id)
    }

    const gitService = new GitService(worktree.path)
    return gitService.listBranches()
  }
}

File: src-tauri/src/services/worktree_service.rs
//! Worktree service for managing git worktrees

use thiserror::Error;
use uuid::Uuid;

use crate::db::{DbPool, WorkspaceRepository, WorktreeRepository};
use crate::services::GitService;
use crate::types::{BranchInfo, GitStatusInfo, UpdateWorktreeInput, Worktree};

#[derive(Error, Debug)]
pub enum WorktreeError {
    #[error("Worktree not found: {0}")]
    NotFound(String),
    #[error("Workspace not found: {0}")]
    WorkspaceNotFound(String),
    #[error("Cannot delete main worktree")]
    CannotDeleteMain,
    #[error("Database error: {0}")]
    Database(String),
    #[error("Git error: {0}")]
    Git(String),
}

pub struct WorktreeService {
    worktree_repo: WorktreeRepository,
    workspace_repo: WorkspaceRepository,
}

impl WorktreeService {
    pub fn new(pool: DbPool) -> Self {
        Self {
            worktree_repo: WorktreeRepository::new(pool.clone()),
            workspace_repo: WorkspaceRepository::new(pool),
        }
    }

    /// List worktrees for a workspace
    pub fn list_worktrees(&self, workspace_id: &str) -> Result<Vec<Worktree>, WorktreeError> {
        self.worktree_repo
            .find_by_workspace_id(workspace_id)
            .map_err(|e| WorktreeError::Database(e.to_string()))
    }

    /// Get a worktree by ID
    pub fn get_worktree(&self, id: &str) -> Result<Worktree, WorktreeError> {
        self.worktree_repo
            .find_by_id(id)
            .map_err(|e| WorktreeError::Database(e.to_string()))?
            .ok_or_else(|| WorktreeError::NotFound(id.to_string()))
    }

    /// Create a new worktree
    pub fn create_worktree(
        &self,
        workspace_id: &str,
        name: &str,
        branch: &str,
        path: Option<&str>,
        create_branch: bool,
    ) -> Result<Worktree, WorktreeError> {
        // Get workspace to get repo path
        let workspace = self
            .workspace_repo
            .find_by_id(workspace_id)
            .map_err(|e| WorktreeError::Database(e.to_string()))?
            .ok_or_else(|| WorktreeError::WorkspaceNotFound(workspace_id.to_string()))?;

        // Determine worktree path
        let worktree_path = path
            .map(|p| p.to_string())
            .unwrap_or_else(|| {
                let parent = std::path::Path::new(&workspace.path)
                    .parent()
                    .unwrap_or(std::path::Path::new("."));
                parent.join(name).to_string_lossy().to_string()
            });

        // Create worktree using git
        let wt_info = GitService::add_worktree(&workspace.path, &worktree_path, branch, create_branch)
            .map_err(|e| WorktreeError::Git(e.to_string()))?;

        // Create database record
        let now = chrono::Utc::now().to_rfc3339();
        let worktree = Worktree {
            id: format!(
                "wt_{}{}",
                chrono::Utc::now().timestamp_millis(),
                &Uuid::new_v4().to_string()[..8]
            ),
            workspace_id: workspace_id.to_string(),
            name: name.to_string(),
            branch: wt_info.branch,
            path: wt_info.path,
            sort_mode: crate::types::SortMode::Free,
            display_order: 0,
            is_main: false,
            created_at: now.clone(),
            updated_at: now,
        };

        let created = self
            .worktree_repo
            .create(&worktree)
            .map_err(|e| WorktreeError::Database(e.to_string()))?;

        // Update workspace counts
        self.workspace_repo
            .update_counts(workspace_id)
            .map_err(|e| WorktreeError::Database(e.to_string()))?;

        Ok(created)
    }

    /// Update a worktree
    pub fn update_worktree(
        &self,
        id: &str,
        input: UpdateWorktreeInput,
    ) -> Result<Worktree, WorktreeError> {
        let mut worktree = self.get_worktree(id)?;

        if let Some(name) = input.name {
            worktree.name = name;
        }
        if let Some(sort_mode) = input.sort_mode {
            worktree.sort_mode = sort_mode;
        }
        if let Some(display_order) = input.display_order {
            worktree.display_order = display_order;
        }

        worktree.updated_at = chrono::Utc::now().to_rfc3339();

        self.worktree_repo
            .update(&worktree)
            .map_err(|e| WorktreeError::Database(e.to_string()))
    }

    /// Delete a worktree
    pub fn delete_worktree(&self, id: &str) -> Result<(), WorktreeError> {
        let worktree = self.get_worktree(id)?;

        if worktree.is_main {
            return Err(WorktreeError::CannotDeleteMain);
        }

        // Get workspace to get repo path
        let workspace = self
            .workspace_repo
            .find_by_id(&worktree.workspace_id)
            .map_err(|e| WorktreeError::Database(e.to_string()))?
            .ok_or_else(|| WorktreeError::WorkspaceNotFound(worktree.workspace_id.clone()))?;

        // Remove worktree from git
        GitService::remove_worktree(&workspace.path, &worktree.path)
            .map_err(|e| WorktreeError::Git(e.to_string()))?;

        // Delete database record
        self.worktree_repo
            .delete(id)
            .map_err(|e| WorktreeError::Database(e.to_string()))?;

        // Update workspace counts
        self.workspace_repo
            .update_counts(&worktree.workspace_id)
            .map_err(|e| WorktreeError::Database(e.to_string()))?;

        Ok(())
    }

    /// Checkout a branch in a worktree
    pub fn checkout_branch(
        &self,
        id: &str,
        branch: &str,
        create: bool,
    ) -> Result<Worktree, WorktreeError> {
        let mut worktree = self.get_worktree(id)?;

        GitService::checkout_branch(&worktree.path, branch, create)
            .map_err(|e| WorktreeError::Git(e.to_string()))?;

        worktree.branch = branch.to_string();
        worktree.updated_at = chrono::Utc::now().to_rfc3339();

        self.worktree_repo
            .update(&worktree)
            .map_err(|e| WorktreeError::Database(e.to_string()))
    }

    /// Reorder worktrees
    pub fn reorder_worktrees(
        &self,
        workspace_id: &str,
        worktree_ids: &[String],
    ) -> Result<Vec<Worktree>, WorktreeError> {
        self.worktree_repo
            .reorder(workspace_id, worktree_ids)
            .map_err(|e| WorktreeError::Database(e.to_string()))?;

        self.list_worktrees(workspace_id)
    }

    /// Get git status for a worktree
    pub fn get_git_status(&self, id: &str) -> Result<GitStatusInfo, WorktreeError> {
        let worktree = self.get_worktree(id)?;
        GitService::get_status(&worktree.path).map_err(|e| WorktreeError::Git(e.to_string()))
    }

    /// List branches for a worktree
    pub fn list_branches(&self, id: &str) -> Result<BranchInfo, WorktreeError> {
        let worktree = self.get_worktree(id)?;
        GitService::list_branches(&worktree.path).map_err(|e| WorktreeError::Git(e.to_string()))
    }
}
